libV2/cache/backend/redis.js

```diff
@@ -0,0 +1,164 @@
+const IORedis = require('ioredis');
+const schema = require('../schema');
+
+const { LoggerContext } = require('../../utils');
+
+const moduleLogger = new LoggerContext({
+    module: 'cache.backend.redis.RedisCache',
+});
+
+class RedisCache {
+    constructor(options, prefix) {
+        this._redis = null;
+        this._options = options;
+        this._prefix = prefix || 'utapi';
+    }
+
+    async connect() {
+        moduleLogger.debug('Connecting to redis...');
+        this._redis = new IORedis(this._options);
+        this._redis
+            .on('error', err =>
+                moduleLogger.error(`error connecting to redis ${err}`))
+            .on('connect', () => moduleLogger.debug('connected to redis'));
+        return true;
+    }
+
+    async disconnect() {
+        const logger = moduleLogger.with({ method: 'disconnect' });
+        if (this._redis) {
+            try {
+                logger.debug('closing connection to redis');
+                await this._redis.quit();
+            } catch (error) {
+                logger.error('error while closing connection to redis', {
+                    error,
+                });
+                throw error;
+            }
+            this._reds = null;
+        } else {
+            logger.debug('disconnect called but no connection to redis found');
+        }
+    }
+
+    async getKey(key) {
+        try {
+            return this._redis.get(key);
+        } catch (error) {
+            moduleLogger
+                .with({ method: 'getKey' })
+                .error('error fetching key from redis', { key });
+            throw error;
+        }
+    }
+
+    async setKey(key, value) {
+        try {
+            const res = await this._redis.set(key, value);
+            return res === 'OK';
+        } catch (error) {
+            moduleLogger
+                .with({ method: 'setKey' })
+                .error('error setting key in redis', { key });
+            throw error;
+        }
+    }
+
+    async addToShard(shard, metric) {
+        const metricKey = schema.getUtapiMetricKey(this._prefix, metric);
+        const shardKey = schema.getShardKey(this._prefix, shard);
+        const logger = moduleLogger.with({ method: 'addToShard' });
+        logger.debug('adding metric to shard', { metricKey, shardKey });
+
+        let results;
+        try {
+            results = await this._redis
```

Comment:
could avoid indexing later by doing...
`let [setResults, saddResults] = await ...`

Reply:
Would also remove the need of the comments later as its more semantic